Feature/disallow dynamic property option

rajyan opened the pull request:
Implemented bool option in `NodeScopeResolver` as mentioned in https://github.com/phpstan/phpstan-src/pull/1223#issuecomment-1103818439 to optionally restore the behavior before https://github.com/phpstan/phpstan-src/pull/1223.
Strictly, it is not exactly the same behavior as before because it disallows dynamic properties in `isset` and `empty` too, but I believe making them consistent is better.

Also, fixes https://github.com/phpstan/phpstan/issues/3171 because it's related.

rajyan committed:
add failing test for issue-3171

rajyan committed:
recursively allow undefined expressions in coalesce, isset, empty

rajyan committed:
add checkDynamicProperties feature

rajyan committed:
change `$currentlyAllowedUndefinedExpressions` to `array<string, true>` because false will not be set anymore

rajyan committed:
cs fix

rajyan commented:
I'll add some tests for the feature.

rajyan committed:
fix duplicate class name

rajyan commented:
It might be better to make this as an option for `AccessPropertiesRule` and `AccessStaticPropertiesRule`?

ondrejmirtes commented:
I worry that this logic needs to be deep in NodeScopeResolver, or not?

If not then yeah, parameter for a rule is nicer.

rajyan committed:
move checkDynamicProperties to AccessPropertiesRule

rajyan committed:
configure checkDynamicProperties in test

rajyan committed:
add tests for checkDynamicProperties

rajyan commented:
> I worry that this logic needs to be deep in NodeScopeResolver, or not?

I think it's possible now. `currentlyAllowedUndefinedExpressions` in NodeScopeResolver is saying that dynamic property inside isset is a valid PHP with no error. I believe it would be valid even after dynamic properties is deprecated/removed after PHP 8.2. `AccessPropertiesRule` can check the `currentlyAllowedUndefinedExpressions` and change the behavior by `$checkDynamicProperties` like this, which I think makes more sense :+1:

rajyan commented:
While this option can optionally restore the behavior before https://github.com/phpstan/phpstan-src/pull/1223#issuecomment-1103818439, it would restore the false positive https://github.com/phpstan/phpstan/issues/3171 too.
It's because `AccessPropertiesRule` is currently treating `Class|null` and dynamic property on `Class` (which are both `$type->canAccessProperties()->maybe()`) in the same way. It's a different issue and I'll fix it separately.

rajyan committed:
`Access to an undefined property` should be appropriate for these cases.
Should be fixed by specifying types in isset.
Related issue https://github.com/phpstan/phpstan/issues/3171

ondrejmirtes commented:
Thank you! Can you please send an update to https://phpstan.org/config-reference#stricter-analysis as well? There's an "Edit this page" link at the bottom :)

I'll update phpstan-strict-rules so that the playground with strict rules + bleedingEdge enabled will have this option on.

ondrejmirtes commented:
Done: https://github.com/phpstan/phpstan-strict-rules/commit/afbe1e3235c51ee3f10ba4c8b32449099016ea02

ondrejmirtes commented:
I'm testing this in the real world and it's really nicely done! The impact is really minimal and exactly in the right weird places where I'd expect it 👍

rajyan commented:
Thanks! I'm glad that I could help solving dynamic properties related issues :smile:
